src/libraries/Microsoft.PowerFx.Core/Public/Engine.cs

```diff
@@ -0,0 +1,144 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.PowerFx.Core;
+using Microsoft.PowerFx.Core.Binding;
+using Microsoft.PowerFx.Core.Glue;
+using Microsoft.PowerFx.Core.Parser;
+using Microsoft.PowerFx.Core.Public;
+using Microsoft.PowerFx.Core.Public.Types;
+using Microsoft.PowerFx.Core.Syntax;
+using Microsoft.PowerFx.Core.Texl.Intellisense;
+using Microsoft.PowerFx.Core.Types;
+
+namespace Microsoft.PowerFx
+{
+    /// <summary>
+    /// Expose binding logic for Power Fx. 
+    /// Derive from this to provide evaluation abilities. 
+    /// </summary>
+    public class Engine : IPowerFxEngine
+    {
+        /// <summary>
+        /// Configuration symbols for this Power Fx engine.
+        /// </summary>
+        public PowerFxConfig Config { get; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="Engine"/> class.
+        /// </summary>
+        /// <param name="powerFxConfig"></param>
+        public Engine(PowerFxConfig powerFxConfig)
+        {
+            powerFxConfig.Lock();
+            Config = powerFxConfig;
+        }
+
+        /// <summary>
+        /// List all functions (both builtin and custom) registered with this evaluator. 
+        /// </summary>
+        public IEnumerable<string> GetAllFunctionNames()
+        {
+            var resolver = CreateResolver(null);
+            return resolver.Functions.Select(func => func.Name);
+        }
+
+        /// <summary>
+        /// Create a resolver for use in binding. This is called from <see cref="Check(string, FormulaType)"/>.
+        /// Base classes can override this is there are additional symbols not in the config.
+        /// </summary>
+        /// <param name="parameterType"></param>
+        /// <returns></returns>
+        private protected virtual SimpleResolver CreateResolver(RecordType parameterType)
```

Comment:
ParamterType arg left over?

Reply:
The derived classes where using this; but per our other discussion, looks like this can be removed now.